Can't compile on Julia 1.7.0-DEV.653 (2021-03-03)

| | |_| | | | (_| |  |  Version 1.7.0-DEV.653 (2021-03-03)
 _/ |\__'_|_|_|\__'_|  |  Commit 502c03974b (0 days old master)
|__/                   |

(@v1.7) pkg> update
    Updating registry at `~/.julia/registries/General`
    Updating git-repo `https://github.com/JuliaRegistries/General.git`
  No Changes to `~/.julia/environments/v1.7/Project.toml`
  No Changes to `~/.julia/environments/v1.7/Manifest.toml`

julia> using Revise
[ Info: Precompiling Revise [295af30f-e4ad-537b-8983-00126c2a3abe]
ERROR: LoadError: LoadError: AssertionError: precompile(Tuple{mbody.sig.parameters[1], Symbol, Bool, Bool, Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, typeof(methods_by_execution!), Any, MI, DocExprs, Module, Expr})
Stacktrace:
  [1] _precompile_()
    @ Revise ~/.julia/packages/Revise/chUcN/src/precompile.jl:54
  [2] top-level scope
    @ ~/.julia/packages/Revise/chUcN/src/packagedef.jl:1417
  [3] include(mod::Module, _path::String)
    @ Base ./Base.jl:386
  [4] include(x::String)
    @ Revise ~/.julia/packages/Revise/chUcN/src/Revise.jl:1
  [5] top-level scope
    @ ~/.julia/packages/Revise/chUcN/src/Revise.jl:20
  [6] include
    @ ./Base.jl:386 [inlined]
  [7] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt64}}, source::Nothing)
    @ Base ./loading.jl:1213
  [8] top-level scope
    @ none:1
  [9] eval
    @ ./boot.jl:369 [inlined]
 [10] eval(x::Expr)
    @ Base.MainInclude ./client.jl:453
 [11] top-level scope
    @ none:1
in expression starting at /home/xijiang/.julia/packages/Revise/chUcN/src/packagedef.jl:1417
in expression starting at /home/xijiang/.julia/packages/Revise/chUcN/src/Revise.jl:1
ERROR: Failed to precompile Revise [295af30f-e4ad-537b-8983-00126c2a3abe] to /home/xijiang/.julia/compiled/v1.7/Revise/jl_0r1bSq.
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] compilecache(pkg::Base.PkgId, path::String, internal_stderr::Base.TTY, internal_stdout::Base.TTY)
   @ Base ./loading.jl:1360
 [3] compilecache(pkg::Base.PkgId, path::String)
   @ Base ./loading.jl:1306
 [4] _require(pkg::Base.PkgId)
   @ Base ./loading.jl:1021
 [5] require(uuidkey::Base.PkgId)
   @ Base ./loading.jl:914
 [6] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:901

--- a/src/precompile.jl
+++ b/src/precompile.jl
@@ -1,82 +1,91 @@
+macro warnpcfail(ex::Expr)
+    modl = __module__
+    file = __source__.file === nothing ? "?" : String(__source__.file)
+    line = __source__.line
+    quote
+        $(esc(ex)) || @warn """precompile directive
+     $($(Expr(:quote, ex)))
+ failed. Please report an issue in $($modl) (after checking for duplicates) or remove this directive.""" _file=$file _line=$line
+    end
+end
+
 function _precompile_()
     ccall(:jl_generating_output, Cint, ()) == 1 || return nothing
 
-    @assert precompile(Tuple{TaskThunk})
-    @assert precompile(Tuple{typeof(wait_changed), String})
-    @assert precompile(Tuple{typeof(watch_package), PkgId})
-    @assert precompile(Tuple{typeof(_watch_package), PkgId})
-    @assert precompile(Tuple{typeof(watch_includes), Module, String})
-    @assert precompile(Tuple{typeof(watch_manifest), String})
-    @assert precompile(Tuple{typeof(revise_dir_queued), String})
-    @assert precompile(Tuple{typeof(revise_file_queued), PkgData, String})
-    @assert precompile(Tuple{typeof(init_watching), PkgData, Vector{String}})
-    @assert precompile(Tuple{typeof(add_revise_deps)})
-    @assert precompile(Tuple{typeof(swap_watch_package), PkgId})
+    @warnpcfail precompile(Tuple{TaskThunk})
+    @warnpcfail precompile(Tuple{typeof(wait_changed), String})
+    @warnpcfail precompile(Tuple{typeof(watch_package), PkgId})
+    @warnpcfail precompile(Tuple{typeof(_watch_package), PkgId})
+    @warnpcfail precompile(Tuple{typeof(watch_includes), Module, String})
+    @warnpcfail precompile(Tuple{typeof(watch_manifest), String})
+    @warnpcfail precompile(Tuple{typeof(revise_dir_queued), String})
+    @warnpcfail precompile(Tuple{typeof(revise_file_queued), PkgData, String})
+    @warnpcfail precompile(Tuple{typeof(init_watching), PkgData, Vector{String}})
+    @warnpcfail precompile(Tuple{typeof(add_revise_deps)})
+    @warnpcfail precompile(Tuple{typeof(swap_watch_package), PkgId})
 
-    @assert precompile(Tuple{typeof(revise)})
-    @assert precompile(Tuple{typeof(revise_first), Expr})
-    @assert precompile(Tuple{typeof(includet), String})
-    @assert precompile(Tuple{typeof(track), Module, String})
+    @warnpcfail precompile(Tuple{typeof(revise)})
+    @warnpcfail precompile(Tuple{typeof(revise_first), Expr})
+    @warnpcfail precompile(Tuple{typeof(includet), String})
+    @warnpcfail precompile(Tuple{typeof(track), Module, String})
     # setindex! doesn't fully precompile, but it's still beneficial to do it
     # (it shaves off a bit of the time)
     # See https://github.com/JuliaLang/julia/pull/31466
-    @assert precompile(Tuple{typeof(setindex!), ExprsSigs, Nothing, RelocatableExpr})
-    @assert precompile(Tuple{typeof(setindex!), ExprsSigs, Vector{Any}, RelocatableExpr})
-    @assert precompile(Tuple{typeof(setindex!), ModuleExprsSigs, ExprsSigs, Module})
-    @assert precompile(Tuple{typeof(setindex!), Dict{PkgId,PkgData}, PkgData, PkgId})
-    @assert precompile(Tuple{Type{WatchList}})
-    @assert precompile(Tuple{typeof(setindex!), Dict{String,WatchList}, WatchList, String})
+    @warnpcfail precompile(Tuple{typeof(setindex!), ExprsSigs, Nothing, RelocatableExpr})
+    @warnpcfail precompile(Tuple{typeof(setindex!), ExprsSigs, Vector{Any}, RelocatableExpr})
+    @warnpcfail precompile(Tuple{typeof(setindex!), ModuleExprsSigs, ExprsSigs, Module})
+    @warnpcfail precompile(Tuple{typeof(setindex!), Dict{PkgId,PkgData}, PkgData, PkgId})
+    @warnpcfail precompile(Tuple{Type{WatchList}})
+    @warnpcfail precompile(Tuple{typeof(setindex!), Dict{String,WatchList}, WatchList, String})
 
     MI = CodeTrackingMethodInfo
-    @assert precompile(Tuple{typeof(minimal_evaluation!), MI, Core.CodeInfo, Symbol})
-    @assert precompile(Tuple{typeof(minimal_evaluation!), Any, MI, Core.CodeInfo, Symbol})
-    @assert precompile(Tuple{typeof(methods_by_execution!), Any, MI, DocExprs, Module, Expr})
-    @assert precompile(Tuple{typeof(methods_by_execution!), Any, MI, DocExprs, JuliaInterpreter.Frame, Vector{Bool}})
-    @assert precompile(Tuple{typeof(Core.kwfunc(methods_by_execution!)),
+    @warnpcfail precompile(Tuple{typeof(minimal_evaluation!), MI, Core.CodeInfo, Symbol})
+    @warnpcfail precompile(Tuple{typeof(minimal_evaluation!), Any, MI, Core.CodeInfo, Symbol})
+    @warnpcfail precompile(Tuple{typeof(methods_by_execution!), Any, MI, DocExprs, Module, Expr})
+    @warnpcfail precompile(Tuple{typeof(methods_by_execution!), Any, MI, DocExprs, JuliaInterpreter.Frame, Vector{Bool}})
+    @warnpcfail precompile(Tuple{typeof(Core.kwfunc(methods_by_execution!)),
                              NamedTuple{(:mode,),Tuple{Symbol}},
                              typeof(methods_by_execution!), Function, MI, DocExprs, Module, Expr})
-    @assert precompile(Tuple{typeof(Core.kwfunc(methods_by_execution!)),
+    @warnpcfail precompile(Tuple{typeof(Core.kwfunc(methods_by_execution!)),
                              NamedTuple{(:skip_include,),Tuple{Bool}},
                              typeof(methods_by_execution!), Function, MI, DocExprs, Module, Expr})
-    @assert precompile(Tuple{typeof(Core.kwfunc(methods_by_execution!)),
+    @warnpcfail precompile(Tuple{typeof(Core.kwfunc(methods_by_execution!)),
                              NamedTuple{(:mode, :skip_include),Tuple{Symbol,Bool}},
                              typeof(methods_by_execution!), Function, MI, DocExprs, Module, Expr})
-    @assert precompile(Tuple{typeof(Core.kwfunc(methods_by_execution!)),
+    @warnpcfail precompile(Tuple{typeof(Core.kwfunc(methods_by_execution!)),
                              NamedTuple{(:mode,),Tuple{Symbol}},
                              typeof(methods_by_execution!), Function, MI, DocExprs, Frame, Vector{Bool}})
-    @assert precompile(Tuple{typeof(Core.kwfunc(methods_by_execution!)),
+    @warnpcfail precompile(Tuple{typeof(Core.kwfunc(methods_by_execution!)),
                              NamedTuple{(:mode, :skip_include),Tuple{Symbol,Bool}},
                              typeof(methods_by_execution!), Function, MI, DocExprs, Frame, Vector{Bool}})
 
     mex = which(methods_by_execution!, (Function, MI, DocExprs, Module, Expr))
     mbody = bodymethod(mex)
-    if Sys.islinux() || Base.VERSION <= v"1.7.0-DEV"
-        @assert precompile(Tuple{mbody.sig.parameters[1], Symbol, Bool, Bool, Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, typeof(methods_by_execution!), Any, MI, DocExprs, Module, Expr})
-        @assert precompile(Tuple{mbody.sig.parameters[1], Symbol, Bool, Bool, Iterators.Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:skip_include,),Tuple{Bool}}}, typeof(methods_by_execution!), Any, MI, DocExprs, Module, Expr})
-    end
+    @warnpcfail precompile(Tuple{mbody.sig.parameters[1], Symbol, Bool, Bool, Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, typeof(methods_by_execution!), Any, MI, DocExprs, Module, Expr})
+    @warnpcfail precompile(Tuple{mbody.sig.parameters[1], Symbol, Bool, Bool, Iterators.Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:skip_include,),Tuple{Bool}}}, typeof(methods_by_execution!), Any, MI, DocExprs, Module, Expr})
     mfr = which(methods_by_execution!, (Function, MI, DocExprs, Frame, Vector{Bool}))
     mbody = bodymethod(mfr)
-    @assert precompile(Tuple{mbody.sig.parameters[1], Symbol, Bool, typeof(methods_by_execution!), Any, MI, DocExprs, Frame, Vector{Bool}})
+    @warnpcfail precompile(Tuple{mbody.sig.parameters[1], Symbol, Bool, typeof(methods_by_execution!), Any, MI, DocExprs, Frame, Vector{Bool}})
 
-    @assert precompile(Tuple{typeof(hastrackedexpr), Expr})
-    @assert precompile(Tuple{typeof(get_def), Method})
-    @assert precompile(Tuple{typeof(parse_pkg_files), PkgId})
+    @warnpcfail precompile(Tuple{typeof(hastrackedexpr), Expr})
+    @warnpcfail precompile(Tuple{typeof(get_def), Method})
+    @warnpcfail precompile(Tuple{typeof(parse_pkg_files), PkgId})
     if isdefined(Revise, :filter_valid_cachefiles)
-        @assert precompile(Tuple{typeof(filter_valid_cachefiles), String, Vector{String}})
+        @warnpcfail precompile(Tuple{typeof(filter_valid_cachefiles), String, Vector{String}})
     end
-    @assert precompile(Tuple{typeof(pkg_fileinfo), PkgId})
-    @assert precompile(Tuple{typeof(push!), WatchList, Pair{String,PkgId}})
-    @assert precompile(Tuple{typeof(pushex!), ExprsSigs, Expr})
-    @assert precompile(Tuple{Type{ModuleExprsSigs}, Module})
-    @assert precompile(Tuple{Type{FileInfo}, Module, String})
-    @assert precompile(Tuple{Type{PkgData}, PkgId})
-    @assert precompile(Tuple{typeof(Base._deleteat!), Vector{Tuple{Module,String,Float64}}, Vector{Int}})
-    @assert precompile(Tuple{typeof(add_require), String, Module, String, String, Expr})
-    @assert precompile(Tuple{typeof(_add_require), String, Module, String, String, Expr})
-    @assert precompile(Tuple{Core.kwftype(typeof(maybe_add_includes_to_pkgdata!)),NamedTuple{(:eval_now,), Tuple{Bool}},typeof(maybe_add_includes_to_pkgdata!),PkgData,String,Vector{Pair{Module, String}}})
+    @warnpcfail precompile(Tuple{typeof(pkg_fileinfo), PkgId})
+    @warnpcfail precompile(Tuple{typeof(push!), WatchList, Pair{String,PkgId}})
+    @warnpcfail precompile(Tuple{typeof(pushex!), ExprsSigs, Expr})
+    @warnpcfail precompile(Tuple{Type{ModuleExprsSigs}, Module})
+    @warnpcfail precompile(Tuple{Type{FileInfo}, Module, String})
+    @warnpcfail precompile(Tuple{Type{PkgData}, PkgId})
+    @warnpcfail precompile(Tuple{typeof(Base._deleteat!), Vector{Tuple{Module,String,Float64}}, Vector{Int}})
+    @warnpcfail precompile(Tuple{typeof(add_require), String, Module, String, String, Expr})
+    @warnpcfail precompile(Tuple{typeof(_add_require), String, Module, String, String, Expr})
+    @warnpcfail precompile(Tuple{Core.kwftype(typeof(maybe_add_includes_to_pkgdata!)),NamedTuple{(:eval_now,), Tuple{Bool}},typeof(maybe_add_includes_to_pkgdata!),PkgData,String,Vector{Pair{Module, String}}})
 
     for TT in (Tuple{Module,Expr}, Tuple{DataType,MethodSummary})
-        @assert precompile(Tuple{Core.kwftype(typeof(Base.CoreLogging.handle_message)),NamedTuple{(:time, :deltainfo), Tuple{Float64, TT}},typeof(Base.CoreLogging.handle_message),ReviseLogger,LogLevel,String,Module,String,Symbol,String,Int})
+        @warnpcfail precompile(Tuple{Core.kwftype(typeof(Base.CoreLogging.handle_message)),NamedTuple{(:time, :deltainfo), Tuple{Float64, TT}},typeof(Base.CoreLogging.handle_message),ReviseLogger,LogLevel,String,Module,String,Symbol,String,Int})
     end
     return nothing
 end

